[WIP] bots: Adding a new 'poll' bot

neiljp opened the pull request:
This is a fairly complete initial implementation of a bot to support polls within streams. It should work against current master, and can often be found running as **Shiny test bot** in #bot testing/neiljp.

I have noted the main design aspects and limitations in the usage text for now; the main limitation I see is currently the parsing of stream and topic names with spaces - if this worked properly, then polls could be started/controlled entirely by private messages, which seems much simpler. This is supported as-is, but spaces in names will confuse it.

Most recently, I added support for one poll per user per topic, rather than just one per topic; this seems sufficient for most cases, though I'm currently seeding the 'id' of the poll with just the user (creator) id.

Reviews and comments gratefully accepted.

smarx commented:
*Automated message from Dropbox CLA bot*

@neiljp, it looks like you've already signed the Dropbox CLA. Thanks!

YagoGG commented on api/bots/poll/poll.py:
```diff
@@ -0,0 +1,266 @@
+                msg = ("Poll created in stream '#%s' with topic '%s':\n"
+                       "%s") % (stream_topic[0], stream_topic[1], new_poll['title'])
+                for i in range(new_poll['n']):
+                    msg += "\n%s" % (new_poll[i])
```
The parenthesis around `new_poll[i]` aren't necessary.

neiljp replied:
Now moved to using .format in the next upload

derAnfaenger commented on api/bots/poll/poll.py:
```diff
@@ -0,0 +1,266 @@
+             - Other vote 'schemes'?
+             - External vote engines?
+             - External vote storage? (current state is 'maintained' in the message too!)
+            '''
```
I think we can move this rather long docstring to the `poll.md` documentation file and only state a rather short example here.

neiljp replied:
I think I mentioned somewhere that this was a temporary measure, brainstorming the current state into the same file. I'll move it now; previously I was just looking for some feedback on whether it looked like a good idea before doing more polish.

neiljp replied:
I've updated the code to include a detailed readme.md, and the bot now has more detailed help commands.

YagoGG commented on api/bots/poll/poll.py:
```diff
@@ -0,0 +1,266 @@
+                           (stream_topic[0], stream_topic[1], poll['title']))
+            for i, tally in enumerate(poll['tallies']):
+                if sender_id in tally:
+                    msg = ("You have already voted" + context_txt +
```
Better insert `context_txt` with `%` as well.

neiljp replied:
Now in format string using .format.

YagoGG commented on api/bots/poll/poll.py:
```diff
@@ -0,0 +1,266 @@
+                poll['tallies'][vote_index-1].append(sender_id)
+                update_poll_text(poll, stream_topic)
+                state_handler.set_state(active_polls)
+                msg = ("You just voted" + context_txt +
```
Same here.

neiljp replied:
Done, into .format.

YagoGG commented on api/bots/poll/poll.py:
```diff
@@ -0,0 +1,266 @@
+            else:
+                private_reply(sender,
+                              ("Please select a number to vote for, between 1-%d,"
+                               % (max_vote_index)) + context_txt)
```
Here as well.

neiljp replied:
Done, into .format.

YagoGG commented on api/bots/poll/poll.py:
```diff
@@ -0,0 +1,266 @@
+            '''
+
+    def handle_message(self, message, bot_handler, state_handler):
+        help_content = "\nCommands: help, new, vote, end"
```
It'd be great if you could add some content from the docstring above (like the usage example) here.

aero31aero replied:
Seconded

neiljp replied:
My amended version (just pushed) has a readme.md with the docstring from before and much improved on-line documentation. Let me know if you agree!

YagoGG commented:
Apparently @derAnfaenger and I have some telepathy with reviewing PRs ;)

Left a couple comments on the code style, @neiljp! The logic looks good to me overall, although we should probably try to reduce the complexity of the code eventually. I guess that the `FIXME`s are what is still in progress, right?

Will try to test it manually in the next few days!

aero31aero commented on api/bots/poll/poll.py:
```diff
@@ -0,0 +1,266 @@
+            vote_options = lines[2:]
+
+        if 0:
+            print("All : %s" % (lines))
```
You might want to use the logger to log these instead of print so that the output could be disabled by passing `--quiet` to `run.py`

neiljp replied:
I've just removed these lines.

neiljp commented:
Thanks for the testing/review everyone!

I've followed up on some comments in the latest amended commit with at least:
- docs moved to readme.md
- added/amended commands: about, help, commands

Other things to look at:
- testing; I had a simple template for this, but the bot got more complex and changed a lot, so I hadn't bothered with it yet.
- logging; I've seen this in use, but need to read up on it more to understand how it works.
- format strings; I understand the push to be consistent, but I recently read about % vs .format, so I'm holding off on those changes until I understand if one is preferred (in python verison x generally, and in zulip server/client/api/bot code)

I also remembered a mention of another PR for something similar, and after searching came up with #2783. I haven't compared features/interface yet, so if anyone knows how that was coming along then or subsequently, that'd be good to know.

neiljp commented:
Corrected command documentation formatting in readme.md.

neiljp commented:
The latest version now supports spaces in stream and topic names using '+' as a space-equivalent.
The documentation has been updated to reflect this.

neiljp commented:
Testing for the straightforward cases now added.
Pending are any tests for what I believe are the currently-unsupported cases of updating messages and one message triggering multiple responses.
Note that the tests as-committed will probably not work, as I have changes to bot_test_lib.py which appear necessary for tests to be possible for the bot, which I hope to put into another PR.

Also, a small change made so that private messages specify the stream/topic first, rather than last, when voting.

neiljp commented:
update_poll_text now returns a tuple including the updated poll message in addition to the success/fail, allowing use in the private messages.

test_poll.py updated to (mostly) fix CI issues.

neiljp commented:
Updated for CI & simplification in test_poll.py to remove boilerplate test code after changes by @derAnfaenger.

neiljp commented:
Switch to .format(), from %, for string formatting, incorporating suggested changes from original reviews.

Remove debug code using print().

neiljp commented:
FYI @YagoGG @derAnfaenger @aero31aero:

Current status:
* Responded to all comments above, both with comments and changes to code/docs/tests.
* Interaction via private messages now fully possible, using '+' in place of spaces
* String formatting improved (and moved to using .format())
* A readme.md now contains operating documentation, including examples and limitations
* More integrated help commands (about/help/commands)
* test_poll.py contains tests for simple commands (about/help/commands/"") and failures which involve no poll state; to run successfully it requires small modifications to bot_test_lib.py, in an upcoming PR
* The only remaining FIXME in poll.py is re handling when update_poll_text fails during voting, which is tied up in what to do in the case of update-message failure

Outstanding external issues open for discussion here or on #integrations:
* Cannot add tests for code which has multiple responses (eg. new command responds privately and in stream) 
* Cannot add tests for code which uses update_message
* No current way for bot to test if update_message will work ahead of time
* External state for bots is not able to be saved via an API (polls die if bot does)

@showell indicated he's not sold on exposing update_message to bots on #general, today. Given the minor 'updates' possible via reactions (and upcoming text-reactions) and thoughts over exposing widgets/canvas features, allowing bots to update messages doesn't seem like something to avoid.

timabbott commented:
Let's plan to have a discussion on chat.zulip.org about the `update_message` question.

neiljp commented:
Closing this, since this is now migrated to zulip/python-zulip-api#41.
